diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -595,99 +616,114 @@ bool DifferentialFunctionComparator::maySkipCast(const User *Cast) const {
+                    // in a different CFG path, fail. We need a single one.
+                    return false;
                 }
-            } else if (isa<StoreInst>(&*it)) {
-                // Check whether a possibly conflicting store instruction
-                // is present. If found, end the search with a failure.
-                PreviousLoad = nullptr;
-                searchPredecessors = false;
-                blockQueue.clear();
+                // Otherwise remember this replacement.
+                Replacement = ReplacementCandidate;
+                checkPredecessors = false;
                 break;
+            } else if (mayStoreToMemoryAtPtr(&*it, Load->getPointerOperand())) {
+                // Fail if a possibly conflicting store or call is found.
+                return false;
             }
         }
 
-        if (searchPredecessors) {
+        if (checkPredecessors) {
             auto BBPredecessors = predecessors(BB);
             if (BBPredecessors.begin() == BBPredecessors.end()) {
-                // If there are no more predecessors available,
-                // end the analysis with a failure.
-                PreviousLoad = nullptr;
-                break;
+                // Fail if no more predecessors are available.
+                return false;
             }
             // Queue up all unvisited predecessors.
             for (auto PredBB : BBPredecessors) {
                 if (visitedBlocks.find(PredBB) == visitedBlocks.end()) {
-                    visitedBlocks.insert(BB);
+                    visitedBlocks.insert(PredBB);
                     blockQueue.push_back(PredBB);
                 }
             }
         }
     }
-    // If the load is repeated without stores in between, skip it because
-    // the load is redundant and its removal will cause no semantic difference.
-    if (PreviousLoad) {
-        replacedInstructions.insert({Load, PreviousLoad});
+    // Successfully found a replacement,
+    if (Replacement) {
+        replacedInstructions.insert({Load, Replacement});
         return true;
     }
     return false;
 }
 
 /// Check whether the given reorderable binary operator can be skipped.
 /// It can only be skipped if all its users are binary operations
-/// of the same kind.
+/// of the same kind or ignorable stores.
 bool DifferentialFunctionComparator::maySkipReorderableBinaryOp(
         const Instruction *Op) const {
-    return std::all_of(Op->user_begin(), Op->user_end(), [Op](auto user) {
-        auto userOp = dyn_cast<BinaryOperator>(user);
-        return userOp && userOp->getOpcode() == Op->getOpcode();
+    return std::all_of(Op->user_begin(), Op->user_end(), [&](auto user) {
+        if (auto userBinOp = dyn_cast<BinaryOperator>(user))
+            return userBinOp->getOpcode() == Op->getOpcode();
+        if (auto userStore = dyn_cast<StoreInst>(user))
+            return hasIgnoredAllocaAsPtrOperand(userStore);
```

Comment:
Why do we need to have this check here? This pattern doesn't seem specific to binary operations, could we come up with some general approach that would allow us to avoid this?

Also this is missing the `config.Patterns.StructChanges` check.

Reply:
Idea: add a method which will follow the stored value and return uses of all loads from the same pointer. Then, check that all uses of these are binary operations.

Reply:
I decided not to follow this approach because I think it's unnecessary. If the result of the skipped binary operation is stored to an ignored variable, we know that the store will always be skipped, which should never lead to false negatives.

I did, however, refactor `maySkipInstruction` to make this more general. See my comment summarizing the changes.